### PR TITLE
#165273471: signup validation, receive descriptive error messages

### PR DESCRIPTION
**Pull request by 3Nakajugo**
#### What does the PR do?
- Validates user signup information and displays descriptive error messages in case of the user inputs invalid data.
#### Description of Task to be completed? validate username
- validate password
- validate email
- validate username
- write unit tests
- update URLs
- update serializers.py
- remove unused import
#### How should this be manually tested?
[signup url](http://127.0.0.1:8000/api/users/)
```
{
  "user":{
  	"username": "nakajugo",
    "email": "example@nexample.com",
    "password": "8789877876uy*"
  }
}
```
#### What are the relevant pivotal tracker stories?
[#165273471](https://www.pivotaltracker.com/story/show/165273471)

#### Screenshot
- register user with an invalid password
image

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -10,12 +9,60 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
 
+    def validate_password(self, password):
+        """
+            validates that  password is longer than 8 characters
+            password is alphanumeric
+        """
+        if len(password) < 8:
+            raise serializers.ValidationError(
+                "Password should atleast be 8 characters.")
+        if not re.search(r'[0-9]', password) or not re.search(r'[a-z]', password):
```
Comment (archibishop) [inline authors/apps/authentication/serializers.py]: @3Nakajugo This seems not check for a special character within the password accepts leading and  trailing spaces in the password. Fix this issue here.
#### Screenshots
*leading spaces with in the password*
Screen Shot 2019-04-29 at 08 25 45
*no special character with in the password*
Screen Shot 2019-04-29 at 08 18 48

File: authors/apps/authentication/tests/test_authentication.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+
+class RegisterAPITestCase(APITestCase):
+    """User manager tests"""
+
+    def setUp(self):
+
+        self.client = APIClient()
+
+        self.user_registration = {
```
Comment (archibishop) [inline authors/apps/authentication/tests/test_authentication.py]: @3Nakajugo Please use fixtures so we can reuse the dummy data in our tests. You could have a look here `https://docs.djangoproject.com/en/2.2/topics/testing/tools/#topics-testing-fixtures`. Tell your colleagues follow a similar approach.

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -10,12 +9,60 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+                "Password should not include white spaces")
+        return password
+
+
+    def validate_username(self, username):
+        """ validates username"""
+        exist_username = User.objects.filter(username=username)
+        if exist_username.exists():
+            raise serializers.ValidationError(
+                "User with the username provided already exists")
```
Comment (archibishop) [inline authors/apps/authentication/serializers.py]:
```suggestion
                "username provided already exists")
```

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -1,7 +1,6 @@
 from django.contrib.auth import authenticate
-
 from rest_framework import serializers
-
+import re
```
Comment (sanya-kenneth) [inline authors/apps/authentication/serializers.py]: Import this after importing the User class
```suggestion
from .models import User
import re
```

**Review by sanya-kenneth — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (3Nakajugo) [reply authors/apps/authentication/serializers.py]: @archibishop  should the special characters be included in the password or exclude them?

Comment (archibishop) [reply authors/apps/authentication/serializers.py]: @3Nakajugo You should include them.

File: authors/apps/authentication/views.py [hunk 5]
```diff
@@ -19,10 +19,11 @@ class RegistrationAPIView(APIView):
 
     def post(self, request):
         user = request.data.get('user', {})
-
+        user['username'] = user['username'].strip()
```
Comment (archibishop) [inline authors/apps/authentication/views.py]: @3Nakajugo Do you think we need a username that has spaces within it.

Comment (archibishop) [reply authors/apps/authentication/views.py]: I think we should only accept a username with no spaces.

**Review by mwinel — CHANGES_REQUESTED** [review]
Please return an appropriate message for a short password.
Please add validations for white spaces at the beginning and the end of fields.
Please validate the username based on characters too. would it be logical to have a username like this one`.....`

**Review by mwinel — APPROVED** [review]

File: authors/apps/authentication/tests/test_authentication.py [hunk 6]
```diff
@@ -97,6 +97,7 @@ def test_user_login(self):
         user.save()
         response = self.client.post(login_url, self.user_test_data.user_login,
                                     format="json")
+        print(response)
```
Comment (archibishop) [inline authors/apps/authentication/tests/test_authentication.py]: @3Nakajugo Remove the print here.

**Comment by archibishop** [comment]
@3Nakajugo Leading whitespaces are still being accepted please look into this. 
Screen Shot 2019-04-30 at 09 49 35

File: authors/apps/authentication/views.py [hunk 7]
```diff
@@ -27,10 +27,10 @@ class RegistrationAPIView(APIView):
 
     def post(self, request):
         user = request.data.get('user', {})
-
         # The create serializer, validate serializer, save serializer pattern
         # below is common and you will see it a lot throughout this course and
         # your own work later on. Get familiar with it.
+
```
Comment (archibishop) [inline authors/apps/authentication/views.py:33]: @3Nakajugo Remove this unnecessary change.

**Review by sanya-kenneth — APPROVED** [review]

**Review by KabohaJeanMark — APPROVED** [review]
Good work on working on the feedback Edna

**Review by archibishop — APPROVED** [review]
@3Nakajugo LGTM I am merging this
